### PR TITLE
fix(embedded): allow guest users to sort table columns in embedded dashboards

### PR DESCRIPTION
**Pull request by rusackas**
### SUMMARY

When viewing an embedded dashboard as a guest user, clicking a table column header to sort raises `Guest user cannot modify chart payload`. This is a false positive from the `query_context_modified()` guard in `superset/security/manager.py`.

**Root cause:** `orderby` was compared with the same strict subset check as `metrics`, `columns`, and `groupby` — the requested `orderby` had to be a subset of the stored chart's `orderby`. Saved charts almost always store `orderby == []`, so any guest sort (e.g. `[["gender", true]]`) fails the subset check and the request is rejected, even though sorting only changes the **ordering** of the result, not which data is read.

**Fix:** `orderby` is now validated separately. A guest may sort by any column or metric the stored chart already references (collected from the chart's `params_dict` and stored `query_context` — columns, groupby, metrics, all_columns, plus existing order-by targets). Any order-by term **not** present in the stored chart — e.g. a free-form `random()` expression — is still rejected, so the existing SQL-injection protection is preserved.

While here, the columns/metrics/groupby comparison is extracted into a small helper (`_columns_metrics_modified`) to keep `query_context_modified` within the complexity budget, and a pre-existing variable-shadowing bug in the inner loop (`for key in equivalent` shadowing the outer `key`) is fixed.

> ⚠️ **Security-sensitive area — please review carefully.** This relaxes the guest payload-tampering guard. The intent is to permit legitimate re-ordering while continuing to reject any order-by term the chart does not already expose. The existing injection test (`test_query_context_modified_orderby`, sorting by `random()`) still passes. cc anyone who worked on the recent `_native_filter_*` hardening in this module — the new check mirrors that `_native_filter_term_allowed` style.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

**Before:** Guest clicks a table column header → `Data error: Guest user cannot modify chart payload`.
**After:** Guest can sort embedded table columns. Sorting by an expression not present in the chart (e.g. `random()`) is still blocked.

### TESTING INSTRUCTIONS

```bash
pytest tests/unit_tests/security/manager_test.py -k "query_context_modified" -v
```

New tests (added alongside the existing `test_query_context_modified_orderby` injection test, which continues to pass):

- `test_query_context_modified_orderby_sort_by_column` — sort by an existing column → allowed
- `test_query_context_modified_orderby_sort_by_metric` — sort by an existing metric → allowed
- `test_query_context_modified_orderby_sort_by_adhoc_metric` — sort by an existing adhoc metric definition → allowed
- `test_query_context_modified_orderby_unknown_column` — sort by a column not in the chart → rejected
- `test_query_context_modified_orderby_empty` — empty order-by → allowed

Manual: open an embedded dashboard with a table chart as a guest user and click a column header to sort — it should reorder instead of erroring.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: superset/security/manager.py [hunk 1]
```diff
@@ -520,42 +520,87 @@ def _native_filter_request_modified(query_context: "QueryContext") -> bool:
...
 
-    # Native-filter data requests have no associated chart (no slice_id). Rather
-    # than accepting any payload, constrain them to the column(s) the dashboard's
-    # native filter is allowed to target; other chartless paths keep prior
-    # behavior (see _native_filter_request_modified).
-    if stored_chart is None:
-        return _native_filter_request_modified(query_context)
+    def add(values: Any) -> None:
+        for value in values or []:
+            allowed.add(freeze_value(value))
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/security/manager.py:542]: These are tiny local closures inside an already-documented helper — a docstring on each feels like noise, so leaving them as-is.

File: superset/security/manager.py [hunk 2]
```diff
@@ -520,42 +520,87 @@ def _native_filter_request_modified(query_context: "QueryContext") -> bool:
...
+        for entry in entries or []:
+            if isinstance(entry, (list, tuple)) and entry:
+                allowed.add(freeze_value(entry[0]))
 
-    # cannot request a different chart
-    if form_data.get("slice_id") != stored_chart.id:
-        return True
+    params = stored_chart.params_dict
+    for key in ("columns", "groupby", "metrics", "all_columns"):
+        add(params.get(key))
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/security/manager.py:551]: Good catch on the legacy singular `metric` — added it to the allowed sort targets and a test for it.

File: superset/security/manager.py [hunk 3]
```diff
@@ -520,42 +520,87 @@ def _native_filter_request_modified(query_context: "QueryContext") -> bool:
...
+    allowed = _collect_sortable_identifiers(stored_chart, stored_query_context)
+    form_data = query_context.form_data or {}
+    requested = list(form_data.get("orderby") or [])
+    for query in query_context.queries:
+        requested.extend(getattr(query, "orderby", None) or [])
+
+    for entry in requested:
+        term = entry[0] if isinstance(entry, (list, tuple)) and entry else entry
+        if freeze_value(term) not in allowed:
+            return True
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/security/manager.py:600]: Fair — order-by entries now have to be `(term, bool)` pairs or they read as tampering, so a bare/nested shape gets rejected before it can blow up query building.

**Review by aminghadersohi — COMMENTED** [review]
## Security review — `query_context_modified` orderby relaxation

**HEAD SHA: 7c1fa4ae725faad65a4d35d19242c5610d48d62a**

The three issues raised by codeant-ai are addressed in the final diff (legacy `metric` singular field added at line 554–555, entry shape validated at lines 584–591, closures left undocumented per author's call). The remainder of this review focuses on what's new.

---

### Automated scan coverage (20 scans, both changed Python files)

| Scan | Result |
|------|--------|
| 1. Inline imports without circular-import guard | 0 violations |
| 2. `any()` without short-circuit concern | 0 violations |
| 3. Mutable default arguments | 0 violations |
| 4. `Optional[X]` vs `X \| None` style | Matches existing file convention |
| 5. `isinstance(x, type) == True` patterns | 0 violations |
| 6. Bare `except` / `except Exception` | 0 violations |
| 7. Hardcoded strings that should be constants | 0 violations (key names are domain model) |
| 8. SQL string interpolation | 0 violations |
| 9. Missing type hints | 0 violations — all three helpers typed |
| 10. Missing return types | 0 violations |
| 11. `print()` statements | 0 violations |
| 12. Dead / unreachable code | 0 violations |
| 13. Overly broad `Any` in signatures | `add(values: Any)` — local closure, acceptable |
| 14. Missing `__init__.py` | N/A |
| 15. N+1 queries | 0 violations — no DB I/O in new code |
| 16. `assert` in production code | 0 violations |
| 17. Magic numbers | 0 violations |
| 18. Hardcoded credentials/secrets | 0 violations |
| 19. Missing edge-case tests | See MEDIUM #2 below |
| 20. Missing docstrings on public functions | 0 violations — all three new helpers documented |

---

### Security analysis

**`freeze_value` / `_strip_overridable_keys`**: `freeze_value` performs deterministic JSON serialization after stripping `timeGrain` from every nested dict/list/tuple (`_strip_overridable_keys` is recursive). Both stored and requested values go through the same call so the comparison is symmetric. Tuple-to-list normalization in nested structures is consistent across both sides. ✓

**SQL injection**: A freeform expression like `random()` serializes to a string not in the stored chart's allowlist and is rejected by `_orderby_modified`. The `freeze_value` comparison requires an exact structural match — approximate or partial strings cannot slip through. ✓

**RLS**: Not modified. Row-level security is applied at query-execution time by the database layer. Sorting the filtered result set cannot bypass it — `ORDER BY` is applied to the rows the guest was already entitled to see. ✓

**Direction bit**: Not validated (any direction is permitted). Correct — direction does not affect which rows are read, only their ordering. Rejecting it would block the primary use case without security benefit. ✓

**Entry shape**: Validated as `(term, bool)` pair at lines 584–591 before `entry[0]` and `entry[1]` are accessed. Malformed shapes (bare string, zero-length list, non-bool second element) are treated as tampering. ✓

**Guest capability boundary**: Guests may sort only by columns/metrics the stored chart already references. They cannot introduce new expressions. This is the minimum relaxation needed and does not widen what data is accessible compared to what the chart was designed to expose. ✓

---

### MEDIUM

#### 1. `_collect_sortable_identifiers` misses `all_columns` from stored query context (`manager.py:562`)

From `params_dict` the function iterates over `("columns", "groupby", "metrics", "all_columns")`. From `stored_query_context.queries` it only iterates over `("columns", "groupby", "metrics")` — `all_columns` is absent:

```python
if stored_query_context:
    for query in stored_query_context.get("queries") or []:
        for key in ("columns", "groupby", "metrics"):   # ← all_columns missing
            add(query.get(key))
        add_orderby(query.get("orderby"))
```

For table charts configured with "show all columns", `QueryObject` writes the exposed columns to `all_columns` in the serialized query context, not `columns`. If the stored chart's `all_columns` in the query context includes columns not present in `params_dict` (e.g., after a chart is resaved with a different column selection), legitimate guest sorts by those columns are silently rejected — the opposite of what this PR intends.

Suggested fix: add `all_columns` to the inner key list at line 562:

```python
for key in ("columns", "groupby", "metrics", "all_columns"):
    add(query.get(key))
```

This is a conservative gap (false positives only — no security regression), but it affects the primary use case.

#### 2. No test independently exercises the stored_query_context path in `_collect_sortable_identifiers` (`manager_test.py:1246–1257`)

Every new test uses `_table_sort_query_context`, which puts the sort column (`"gender"`) in `params_dict.groupby`. The stored query context entry (`columns: ["gender"]`) is therefore redundant — removing the entire stored_query_context branch of `_collect_sortable_identifiers` would leave all the new positive tests green.

A focused test should verify that a column appearing **only** in `stored_query_context.queries[].columns` (not in `params_dict`) is accepted:

```python
def test_query_context_modified_orderby_sort_by_stored_qc_only_column(mocker):
    """A column present only in the stored query context is an allowed sort target."""
    query_context = _table_sort_query_context(mocker, orderby=[("age", True)])
    # "age" not in params_dict, but present in the stored query context
    query_context.slice_.params_dict = {"groupby": ["gender"], "metrics": ["count"]}
    query_context.slice_.query_context = json.dumps(
        {"queries": [{"columns": ["gender", "age"], "metrics": ["count"]}]}
    )
    assert not query_context_modified(query_context)
```

Without this the stored_query_context branch has no independent coverage; a silent break there would go undetected by the current suite.

---

### LOW

#### 3. `_orderby_modified` collects the same entry twice when frontend populates both sources (`manager.py:580–584`)

```python
requested = list(form_data.get("orderby") or [])
for query in query_context.queries:
    requested.extend(getattr(query, "orderby", None) or [])
```

In the common frontend path both `form_data["orderby"]` and `QueryObject.orderby` carry the same sort, so each entry is validated twice in the loop below. This is harmless for correctness (the result is the same) but can confuse a future reader who wonders whether the union is intentional or an oversight. A brief comment explaining that both sources must be validated — because either can carry unauthorized terms — would resolve the ambiguity.

---

### NIT

#### 4. Asymmetric `stored_values` in `_columns_metrics_modified` lacks an explanatory comment (`manager.py:616–636`)

The two subset checks inside the loop use `stored_values` at different stages of construction:

```python
stored_values = {freeze_value(v) for v in stored_chart.params_dict.get(key) or []}
if not requested_values.issubset(stored_values):  # params_dict only
    return True
...
if stored_query_context:
    stored_values.update(...)                      # augmented here
if not queries_values.issubset(stored_values):     # augmented stored_values
    return True
```

`requested_values` (from `form_data`) is checked against params_dict alone; `queries_values` (from `QueryObject`) gets the fuller stored_query_context view. This was the pre-existing behavior, not a regression introduced here, but it is now more visible in the extracted helper. A one-line comment noting the intentional asymmetry would help future reviewers understand why the two checks differ.

**Comment by rusackas** [comment]
@aminghadersohi good catches — pushed a fix:

- MEDIUM #1: added `all_columns` to the stored-query-context loop in `_collect_sortable_identifiers`, with a regression test that goes red without it.
- MEDIUM #2: that test exercises the stored-qc branch in isolation (column present only in the query context, not `params_dict`).
- LOW #3 / NIT #4: added the explanatory comments on the order-by union and the intentional `stored_values` asymmetry.

Left the direction-bit and closure-docstring calls as-is per the earlier reasoning.

**Review by aminghadersohi — APPROVED** [review]
All four prior findings addressed at the new HEAD: `all_columns` added to the stored-query-context collection path in `_collect_sortable_identifiers`, stored-QC-only column test added (`test_query_context_modified_orderby_sort_by_stored_qc_only_column`), double-counting comment added to `_orderby_modified`, and asymmetric `stored_values` comment added to `_columns_metrics_modified`. 20 scans clean. All 4 review threads resolved. CI fully green.
